db/skipped_sequence.go

```diff
@@ -0,0 +1,211 @@
+		timeStamp := v.getTimestamp()
+		if (timeNow - timeStamp) >= maxWait {
+			indexToDelete++
+		} else {
+			// exit early to avoid iterating through whole slice
+			break
+		}
+	}
+	// if we have incremented the index counter, we need to the index range from the slice
+	if indexToDelete > -1 {
+		s.list = slices.Delete(s.list, 0, indexToDelete+1)
+	}
+	// resize slice to reclaim memory if we need to
+	s.clip(ctx)
+}
+
+// clip will clip the capacity of the slice to a constant value if the current capacity of the slice in 2.5x the length
+func (s *SkippedSequenceSlice) clip(ctx context.Context) {
+	// threshold is 2.5x the current length of the slice
+	threshold := 2.5 * float64(len(s.list))
+
+	if cap(s.list) > int(threshold) {
+		// check if we can safely clip without an out of bound errors
+		if (len(s.list) + ClipCapacityHeadroom) < cap(s.list) {
+			base.DebugfCtx(ctx, base.KeyCache, "clipping skipped list capacity")
+			s.list = s.list[:len(s.list):ClipCapacityHeadroom]
+		}
+	}
+}
+
+// findSequence will use binary search to search the elements in the slice for a given sequence
+func (s *SkippedSequenceSlice) findSequence(x uint64) (int, bool) {
+	i, found := slices.BinarySearchFunc(s.list, x, func(a SkippedSequenceListInterface, seq uint64) int {
+		singleSeq, ok := a.(*SingleSkippedSequence)
+		if ok {
+			if singleSeq.seq > seq {
+				return 1
+			}
+			if singleSeq.seq == seq {
+				return 0
+			}
+			return -1
+		}
+		// should never get here as it stands, will have extra handling here pending CBG-3853
+		return 1
+	})
+	return i, found
+}
+
+// removeSeq will remove a given sequence from the slice if it exists
+func (s *SkippedSequenceSlice) removeSeq(x uint64) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	index, found := s.findSequence(x)
+	if !found {
+		return fmt.Errorf("sequence %d not found in the skipped list", x)
+	}
+	// handle boarder cases where x is equal to start or end sequence on range (or just sequence for single entries)
+	if !s.list[index].isRange() {
+		// if not a range, we can just remove the single entry from the slice
+		if s.list[index].getLastSeq() == x {
+			s.list = slices.Delete(s.list, index, index+1)
+		}
+		return nil
+	}
+	// more range handling here CBG-3853, temporarily error as we shouldn't get to this point at this time
+	return fmt.Errorf("entered range handling code")
+}
+
+// insert will insert element in middle of slice maintaining order of rest of slice
+func (s *SkippedSequenceSlice) insert(index int, entry SkippedSequenceListInterface) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.list = append(s.list[:index+1], s.list[index:]...)
+	s.list[index] = entry
+}
+
+// PushSkippedSequenceEntry will append a new skipped sequence entry to the end of the slice, if adding a contiguous
+// sequence function will expand the last entry of the slice to reflect this
+// nolint:staticcheck
+func (s *SkippedSequenceSlice) PushSkippedSequenceEntry(entry SkippedSequenceListInterface) {
```

Comment:
Until we support ranges, I think defining this as 
```suggestion
func (s *SkippedSequenceSlice) PushSkippedSequenceEntry(entry *SingleSkippedSequence) {
```
makes more sense.  Then we can make the decision in CBG-3853 about whether it makes sense to use this function when adding ranges, or if ranges should get a separate function like PushSkippedSequenceRange